library/Freckle/App/Database.hs

```diff
@@ -65,12 +73,68 @@ makePostgresPool = do
   makePostgresPoolWith conf
 
 runDB
-  :: (HasSqlPool app, MonadUnliftIO m, MonadReader app m)
+  :: ( HasActiveConnectionGauge app
+     , HasSqlPool app
+     , Stats.HasStatsClient app
+     , MonadHandler m
+     , MonadUnliftIO m
+     , MonadReader app m
+     )
   => SqlPersistT m a
   -> m a
 runDB action = do
+  app <- ask
   pool <- asks getSqlPool
-  runSqlPool action pool
+  mVaultData <- vaultDataFromRequest <$> waiRequest
+  maybe
+    runSqlPool
+    (runSqlPoolXRay "runDB")
+    mVaultData
+    (bracket_ (increment app) (decrement app) action)
+    pool
+  where
+    increment = withGauge Gauge.increment
+    decrement = withGauge Gauge.decrement
+    withGauge f = runReaderT $ f =<< asks getActiveConnectionGauge
+
+-- | @'runSqlPool'@ but with XRay tracing
+--
+-- Arguments are ordered to promote composition:
+--
+-- 1. Name before vault data, since Vault data could be @'Maybe'@
+-- 2. Action than pool, to match @'runSqlPool'@
+--
+-- Together, this leads to the nice composition:
+--
+-- @
+-- mVaultData <- getVaultData
+-- maybe 'runSqlPool' ('runSqlPoolXRay' "name") mVaultData action pool
+-- @
+--
+-- N.B. This could make sense in an aws-xray-client-persistent package.
+--
+runSqlPoolXRay
```

Comment:
Similar question on `Xray`. Should we expect all our apps to have xray? Does this fail gracefully? What happens when `Xray` doesn't exist?

Reply:
Both EKG and Xray were called out as desired everywhere in the task. @pbrisbin

Reply:
It all should be opt-in.

In this case, I believe `vaultDataFromRequest` is the magic sauce. If the end-user [hasn't done what's required](https://github.com/freckle/megarepo/blob/f701070c8b1ee878fd4afe32d9be0dbec90e41d1/backend/fancy-api/library/Application.hs#L336) for that to work, you hit the `runSqlPool` case. The docs could probably make that clearer, that the WAI Middleware is required before in-band traces (i.e. this `runDB`) will function.